### PR TITLE
Add collectors for ASP.NET Server and Applications and MS data client for Sql

### PR DESCRIPTION
**Pull request by jaco-van-zyl-derivco**
Adding collectors for ASP.NET and Microsoft Data Client metrics

## asp_net_system
```
# HELP windows_asp_net_system_application_restarts The number of times that an application has been restarted during the Web server's lifetime.
# TYPE windows_asp_net_system_application_restarts counter
windows_asp_net_system_application_restarts 0
# HELP windows_asp_net_system_applications_running The number of applications running concurrently on the server computer.
# TYPE windows_asp_net_system_applications_running gauge
windows_asp_net_system_applications_running 0
# HELP windows_asp_net_system_request_wait_time The number of milliseconds that the most recent request waited in the queue for processing.
# TYPE windows_asp_net_system_request_wait_time gauge
windows_asp_net_system_request_wait_time 0
# HELP windows_asp_net_system_requests_disconnected The number of requests that have been disconnected due to a communication failure.
# TYPE windows_asp_net_system_requests_disconnected counter
windows_asp_net_system_requests_disconnected 0
# HELP windows_asp_net_system_requests_queued The number of requests waiting for service from the queue. When this number starts to increment linearly with increased client load, the Web server computer has reached the limit of concurrent requests that it can process.
# TYPE windows_asp_net_system_requests_queued gauge
windows_asp_net_system_requests_queued 0
# HELP windows_asp_net_system_requests_rejected The total number of requests not executed because of insufficient server resources to process them.
# TYPE windows_asp_net_system_requests_rejected gauge
windows_asp_net_system_requests_rejected 0
# HELP windows_asp_net_system_worker_process_restarts The number of times a worker process has been restarted on the server computer.
# TYPE windows_asp_net_system_worker_process_restarts counter
windows_asp_net_system_worker_process_restarts 0
# HELP windows_asp_net_system_worker_processes_running The number of worker processes running on the server computer.
# TYPE windows_asp_net_system_worker_processes_running gauge
windows_asp_net_system_worker_processes_running 0
```

## asp_net_application
```
# HELP windows_asp_net_application_errors_during_execution The total number of errors that occur during the execution of an HTTP request, excluding parser and compilation errors.
# TYPE windows_asp_net_application_errors_during_execution gauge
windows_asp_net_application_errors_during_execution 0
# HELP windows_asp_net_application_errors_total_per_sec The number of errors per second that occur during the execution of HTTP requests, including any parser, compilation, or run-time errors.
# TYPE windows_asp_net_application_errors_total_per_sec counter
windows_asp_net_application_errors_total_per_sec 0
# HELP windows_asp_net_application_managed_memory_used_estimated Estimated memory usage for the application
# TYPE windows_asp_net_application_managed_memory_used_estimated gauge
windows_asp_net_application_managed_memory_used_estimated 0
# HELP windows_asp_net_application_percent_managed_processor_time_estimated Estimated cpu usage for the application
# TYPE windows_asp_net_application_percent_managed_processor_time_estimated gauge
windows_asp_net_application_percent_managed_processor_time_estimated 0
# HELP windows_asp_net_application_request_wait_time_ms The number of milliseconds the most recent request waited in the queue before being processed
# TYPE windows_asp_net_application_request_wait_time_ms gauge
windows_asp_net_application_request_wait_time_ms 0
# HELP windows_asp_net_application_requests_executing The number of requests currently executing.
# TYPE windows_asp_net_application_requests_executing gauge
windows_asp_net_application_requests_executing 0
# HELP windows_asp_net_application_requests_in_application_queue The number of requests currently in the queue.
# TYPE windows_asp_net_application_requests_in_application_queue counter
windows_asp_net_application_requests_in_application_queue 0
# HELP windows_asp_net_application_requests_per_sec The number of requests executed per second.
# TYPE windows_asp_net_application_requests_per_sec counter
windows_asp_net_application_requests_per_sec 0
```

## net_dataprovider_sqlserver
```
# HELP windows_net_dataprovider_sqlserver_hard_connects_per_second The number of connections per second that are being made to a database server.
# TYPE windows_net_dataprovider_sqlserver_hard_connects_per_second gauge
windows_net_dataprovider_sqlserver_hard_connects_per_second{process="process_name[pid]"} 13
# HELP windows_net_dataprovider_sqlserver_hard_disconnects_per_second The number of disconnects per second that are being made to a database server.
# TYPE windows_net_dataprovider_sqlserver_hard_disconnects_per_second gauge
windows_net_dataprovider_sqlserver_hard_disconnects_per_second{process="process_name[pid]"} 12
# HELP windows_net_dataprovider_sqlserver_number_of_active_connection_pool_groups The number of unique connection pool groups that are active. This counter is controlled by the number of unique connection strings that are found in the AppDomain.
# TYPE windows_net_dataprovider_sqlserver_number_of_active_connection_pool_groups gauge
windows_net_dataprovider_sqlserver_number_of_active_connection_pool_groups{process="process_name[pid]"} 1
# HELP windows_net_dataprovider_sqlserver_number_of_active_connection_pools The total number of connection pools.
# TYPE windows_net_dataprovider_sqlserver_number_of_active_connection_pools gauge
windows_net_dataprovider_sqlserver_number_of_active_connection_pools{process="process_name[pid]"} 1
# HELP windows_net_dataprovider_sqlserver_number_of_active_connections The number of active connections that are currently in use.
# TYPE windows_net_dataprovider_sqlserver_number_of_active_connections gauge
windows_net_dataprovider_sqlserver_number_of_active_connections{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_number_of_free_connections The number of connections available for use in the connection pools.
# TYPE windows_net_dataprovider_sqlserver_number_of_free_connections gauge
windows_net_dataprovider_sqlserver_number_of_free_connections{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_number_of_inactive_connection_pool_groups The number of unique connection pool groups that are marked for pruning.
# TYPE windows_net_dataprovider_sqlserver_number_of_inactive_connection_pool_groups gauge
windows_net_dataprovider_sqlserver_number_of_inactive_connection_pool_groups{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_number_of_inactive_connection_pools The number of inactive connection pools that have not had any recent activity and are waiting to be disposed.
# TYPE windows_net_dataprovider_sqlserver_number_of_inactive_connection_pools gauge
windows_net_dataprovider_sqlserver_number_of_inactive_connection_pools{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_number_of_non_pooled_connections The number of active connections that are not pooled.
# TYPE windows_net_dataprovider_sqlserver_number_of_non_pooled_connections gauge
windows_net_dataprovider_sqlserver_number_of_non_pooled_connections{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_number_of_pooled_connections The number of active connections that are being managed by the connection pooling infrastructure.
# TYPE windows_net_dataprovider_sqlserver_number_of_pooled_connections gauge
windows_net_dataprovider_sqlserver_number_of_pooled_connections{process="process_name[pid]"} 1
# HELP windows_net_dataprovider_sqlserver_number_of_reclaimed_connections The number of connections that have been reclaimed through garbage collection where Close or Dispose was not called by the application.
# TYPE windows_net_dataprovider_sqlserver_number_of_reclaimed_connections gauge
windows_net_dataprovider_sqlserver_number_of_reclaimed_connections{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_number_of_stasis_connections The number of connections currently awaiting completion of an action and which are therefore unavailable for use by your application.
# TYPE windows_net_dataprovider_sqlserver_number_of_stasis_connections gauge
windows_net_dataprovider_sqlserver_number_of_stasis_connections{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_soft_connects_per_second The number of active connections being pulled from the connection pool.
# TYPE windows_net_dataprovider_sqlserver_soft_connects_per_second gauge
windows_net_dataprovider_sqlserver_soft_connects_per_second{process="process_name[pid]"} 0
# HELP windows_net_dataprovider_sqlserver_soft_disconnects_per_second The number of active connections that are being returned to the connection pool.
# TYPE windows_net_dataprovider_sqlserver_soft_disconnects_per_second gauge
windows_net_dataprovider_sqlserver_soft_disconnects_per_second{process="process_name[pid]"} 0
```

### CODE REVIEW

**Comment by DavidSkDerivco** [comment]
@breed808 are you the right person to help bump this?

File: collector/asp_net_application.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+import (
+	"github.com/prometheus-community/windows_exporter/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/yusufpapurcu/wmi"
+)
+
+func init() {
+	registerCollector("asp_net_application", newasp_net_applicationCollector) // TODO: Add any perflib dependencies here
+}
```
Comment (breed808) [inline collector/asp_net_application.go:11]: You'll want to remove any `init()` functions in the collectors in favour of the `collectors` variable in `collector/init.go`.

If there's no perflib dependencies you can remove the TODO comment.

File: collector/asp_net_application.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	const subsystem = "asp_net_application"
+	return &asp_net_applicationCollector{
+		ErrorsDuringExecution: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "errors_during_execution"),
+			"The total number of errors that occur during the execution of an HTTP request, excluding parser and compilation errors.",
+			nil,
+			nil,
+		),
+		ErrorsTotalPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "errors_total_per_sec"),
```
Comment (breed808) [inline collector/asp_net_application.go:35]: If this is a counter it'll need a `_total` suffix. Description will need to updated to reflect the metric nature, being the total number of errors.

File: collector/asp_net_application.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		c.RequestsExecuting,
+		prometheus.GaugeValue,
+		float64(dst[0].RequestsExecuting),
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.RequestsInApplicationQueue,
+		prometheus.CounterValue,
+		float64(dst[0].RequestsInApplicationQueue),
+	)
```
Comment (breed808) [inline collector/asp_net_application.go:145]: Should the metric type be Counter? The metric description indicates it's the number of requests currently in the queue, which implies a Gauge metric.

File: collector/asp_net_application.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			"The number of requests currently in the queue.",
+			nil,
+			nil,
+		),
+		RequestsPerSec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "requests_per_sec"),
+			"The number of requests executed per second.",
+			nil,
+			nil,
+		),
```
Comment (breed808) [inline collector/asp_net_application.go:69]: Similar situation here: if it's a counter metric it'll need the `_total` suffix and an updated metric description.

**Review by breed808 — CHANGES_REQUESTED** [review]
Thanks for submitting this! I've left some comments for the `asp_net_application` collector, which also apply to the other collectors introduced in this pull request. Have a look at the comments and let me know if you have any questions.

It'll be necessary to  rebase the feature branch on `master` to remove the duplicate dependabot commits.

**Pull request closed without merging**